### ISSUE TITLE
Handle usage and creation of certs inside the CLI

### ISSUE THREAD

**Issue by ahmetalpbalkan** [Azure/azure-xplat-cli#1710]
As a beginner user I don't see a reason why SSH related params (e.g. `--ssh-cert`) passed to `azure vm create` are needed.

We have couple of options here:

- Check for user's `~/.ssh/id_rsa.pub` (that's where 90% of the people's ssh keys live) Needs  #1681.
- Create a SSH key and store them at `~/.azure` with correct permission and associate them with VMs (what several other cloud providers' CLIs do) –and have a command like `azure vm ssh` to wrap ssh command.

This will eliminate credentials from `azure vm create` and will eventually make that cmd shorter. 

cc: @kmouss @coreysa

**Comment by devigned** [comment]
:+1:

**Comment by polarapfel** [comment]
I have to disagree. No tool should automatically assume the default key (id_rsa.pub) is the desired key. Hence, a parameter to specify the desired key is still required. What needs to be improved though is the support for standard RSA keys as generated by ssh-keygen. Wrapping keys in a x509 certificate through openssl is ridiculous.

**Comment by ahmetalpbalkan** [comment]
@polarapfel RSA format is a separate issue (#1681). We can create and store ssh keys for the user and let them ssh via our cli and we can handle the rest for them, that's what I'm proposing here. This can be a default option but users should always be able to specify the SSH key path if they like.

**Comment by devigned** [comment]
I've always been a fan of the simplicity of using vagrant to ssh into a vm. 

Perhaps, if a user doesn't add the --ssh-cert value, we prompt them and enumerate if we see any ~/.ssh/*.pub?

**Comment by ahmetalpbalkan** [comment]
@devigned that also sounds doable (I think the assumption that ~/.ssh is where keys are stored is correct) and they still have a way to override --ssh-cert.

Alternatively what I'm proposing here is let us handle all the SSH certs if user doesn't provide any SSH certs at all. When user runs `azure vm create` without  --ssh-cert, we create a key, save it  to `~/.azure/certs`, associate the cert with the VM name, create a command like `azure vm ssh` so that user can directly ssh to the VM (we figure out what key to use) and user can have an end-to-end experience without worrying about ssh keys at all.

**Comment by devigned** [comment]
@ahmetalpbalkan, I like that idea for a couple reasons. 

1. No passwords
2. Simple for the user

I'm hesitant because it'll be magic... I wonder how many users will ask where did we hide their key when they try to ssh with out the azure cli.

**Comment by squillace** [comment]
+1: @ahmetalpbalkan @devigned  but prompt user for what you're going to use, and where you might store something if you have it. On the other hand, for scripting scenario, you'd need to have switches to pass desires so that prompt does not appear, including if you have id_rsa but you don't WANT that one and instead want to create a new set. It's not magic if you make it clear what you're doing, and what happens if they say no. Also, the "if they say no" path needs to be very graceful, I should think. FWIW.

**Comment by ruslangabitov** [comment]
Added creation of SSH keys inside the 'vm create' and 'vm docker create' commands: https://github.com/MSOpenTech/azure-xplat-cli/pull/391

Usage: 
  Command creates certs for each VM and saves them in <code>~/.azure/</code> folder or uses them if they are already there.
![image](https://cloud.githubusercontent.com/assets/6327511/8759494/c58a0a3e-2ca9-11e5-987f-870717204d98.png)

  To generate certs you should specify <code>--generate-ssh-keys</code> option.
  For example: <code>azure vm create *resource-group* *vm-name* *location* *OS* --generate-ssh-keys</code>
  If  <code>--ssh-cert </code> (in ASM) or  <code>-ssh-publickey-file </code> (in ARM) are set <code>--generate-ssh-keys</code> will be ignored.

 After keys are created command outputs private key's location:
![image](https://cloud.githubusercontent.com/assets/6327511/8759243/ab2cfd7c-2ca4-11e5-9244-3602013425cb.png)

**Comment by ahmetalpbalkan** [comment]
@ruslangabitov this is pretty cool! Does it overwrite for the new VMs created with the same name?

**Comment by ruslangabitov** [comment]
@ahmetalpbalkan no, it will use certs if there are already there. 
yes, it saves them directly under <code>~/.azure</code>.

Yeah, I'll try to add 'vm ssh' command next, just wanted to share this part of the implementation first.

**Comment by ahmetalpbalkan** [comment]
@ruslangabitov this is great! can't wait to try the next steps as well. I think we should put the keys into `~/.azure/keys`, the root dir might be hard to deal with after some point. :-)

**Comment by ruslangabitov** [comment]
@ahmetalpbalkan yeah, good point, I'll change the folder.

**Comment by sreekanthbs** [comment]
@ahmetalpbalkan Now vm create and vm docker create generates SSh keys inside (~/.azure/ssh/ ) folder in https://github.com/MSOpenTech/azure-xplat-cli/pull/442/
![1710_442_2](https://cloud.githubusercontent.com/assets/11171941/8795542/76d702f2-2fab-11e5-882b-b4f03bb15192.PNG)

![1710_442](https://cloud.githubusercontent.com/assets/11171941/8795530/56c29422-2fab-11e5-9f2a-258b3de2c28d.PNG)

 ![1710_442_3](https://cloud.githubusercontent.com/assets/11171941/8795547/8a3e7956-2fab-11e5-993c-c47a3a835cf4.PNG)

**Comment by ahmetalpbalkan** [comment]
Great stuff.. @ruslangabitov do you have any new we goodness can try out here?

**Comment by squillace** [comment]
@ruslangabitov when do we get this?

**Comment by squillace** [comment]
@ruslangabitov @ahmetalpbalkan I note in https://github.com/Azure/azure-xplat-cli/issues/1710#issuecomment-123204193 that we are STILL specifying .pem files in testing (bad; they include private information). Here's the help for `azure vm create` in arm right now:

```
virtual-machines [sshtypo] % azure vm create --help
help:    Create a VM
help:
help:    Usage: vm create [options] <dns-name> <image> [userName] [password]
<snip />
help:      -t, --ssh-cert <openssh-rsa-file|pem-file|fingerprint>  the SSH certificate
```

No one should use .pem files as certificate files because they contain private information. We have heard this, and we now take ssh rsa format .pub files! Yay! BUT: here's the `azure vm docker create` command. Why are we still taking pem files for the ssh-key in  `docker create vm` command? 

```
azure vm docker create --help
<snip />
help:      -t, --ssh-cert <pem-file|fingerprint>  the SSH certificate
```

can we have unity? parallelism? can we update all creation commands at one and the same time? Should I file a separate issue there?

**Comment by ahmetalpbalkan** [comment]
@squillace +1. But I think we should open a separate issue to discuss/track this. I feel like this'll get lost here.

**Comment by squillace** [comment]
done. https://github.com/Azure/azure-xplat-cli/issues/2393

**Comment by huangpf** [comment]
@squillace If the issues in this thread have been resolved, can we close this one, and focus on #2393? Thanks.

**Comment by ahmetalpbalkan** [comment]
@huangpf this issue should not be closed. The title says "handle usage and creation of VM SSH certs inside the CLI". I think we're not there yet and this thread must discuss it.

In other words, this thread is not about various commandlets not taking the --ssh-key argument; it is about hiding the SSH key management entirely from the user by default. Please read the title and the initial issue description.

**Comment by huangpf** [comment]
Then the label says ```Ready For Test``` should probably be removed.

**Comment by huangpf** [comment]
This should be a very convenient feature for Linux CLI users, but I don't think we can "eliminate credentials from ```azure vm create```..." because CLI can also be used to create Windows VMs by providing user name & password.

**Comment by ahmetalpbalkan** [comment]
@huangpf yeah but for Linux, username/password is not a reality. I'm not saying we should eliminate it for Windows, simply because SSH is obviously for Linux. If you haven't already, please download gcloud cli and create some instances to understand this proposal.

**Comment by huangpf** [comment]
@ahmetalpbalkan Thanks for the suggestion again. We will look into it. @sauryadas.

**Comment by ahmetalpbalkan** [comment]
![image](https://cloud.githubusercontent.com/assets/159209/16441938/3a7e6516-3d83-11e6-9f22-0a86ec1c3278.png)
